src/async-snowplow.ts

```diff
@@ -0,0 +1,119 @@
+export type SnowplowFn = (...args: any[]) => void;
+
+/**
+ * Wraps Snowplow method and allows for flushing a bunch of Snowplow calls.
+ */
+export interface AsyncSnowplow {
+  /**
+   * Same as calling Snowplow but will queue the events if Snowplow is not loaded.
+   */
+  call(...args: any[]): void;
+
+  /**
+   * Tries to flush the queue if Snowplow is setup.
+   */
+  tryFlush(): void;
+}
+
+/**
+ * A version of AsyncSnowplow that runs the calls immediately.  A valid
+ * SnowplowFn must be passed into the constructor.
+ */
+export class ImmediateAsyncSnowplow implements AsyncSnowplow {
+  snowplow: SnowplowFn;
+  handleLogError: (err: Error) => void;
+
+  public constructor(snowplow: SnowplowFn, handleLogError: (err: Error) => void) {
+    this.snowplow = snowplow;
+    this.handleLogError = handleLogError;
+  }
+
+  call(...args: any[]) {
+    try {
+      this.snowplow(...args);
+    } catch (e) {
+      this.handleLogError(e);
+    }
+  }
+
+  tryFlush() {
+    // Do nothing.
+  }
+}
+
+// @ts-expect-error window does not have snowplow on it.
+export const windowSnowplowProvider = () => typeof window !== 'undefined' && window?.snowplow;
+
+const SLEEP_DURATION_MS = 2000;
+const MAX_TIMER_ATTEMPTS = 15;
+
+/**
+ * An implementation of AsyncSnowplow that is backed by a timer.
+ */
+export class TimerAsyncSnowplow implements AsyncSnowplow {
+  snowplowProvider: () => SnowplowFn | undefined;
+  handleLogError: (err: Error) => void;
+  queue: any[][];
+  timer: ReturnType<typeof setTimeout> | undefined;
+  numTimerAttempts = 0;
+
+  public constructor(snowplowProvider: () => SnowplowFn | undefined, handleLogError: (err: Error) => void) {
+    this.snowplowProvider = snowplowProvider;
+    this.handleLogError = handleLogError;
+    this.queue = [];
+  }
+
+  call(...args: any[]) {
+    const snowplow = this.snowplowProvider();
+    if (snowplow) {
+      snowplow(...args);
+    } else {
+      // Delay call.
+      this.queue.push(args);
+      this.createTimerIfNeeded();
+    }
+  }
+
+  tryFlush() {
+    if (this.queue.length > 0) {
+      try {
+        const snowplow = this.snowplowProvider();
+        if (snowplow) {
+          const newQueue: any[][] = [];
+          let error: Error | undefined = undefined;
+          this.queue.forEach((args) => {
+            try {
+              snowplow(...args);
+            } catch (e) {
+              newQueue.push(args);
```

Comment:
Is there a way that bad input would cause events to continually get re-queued?

Reply:
Not with how it's currently coded up.  Most errors are handled internally in Snowplow tracker (e.g. retrying failed RPCs).

With how it's currently coded up, `tryFlush` will stop getting called once Snowplow has loaded.  Putting the failed events back into `queue` is pretty similar to dropping it.

I'd be fine dropping the events altogether if the call to Snowplow fails.  This reduces the risk and doesn't change the current impact of the code?

Thoughts?  Might be easier to VC through this.

Reply:
Gotcha. Sounds reasonable. 

On a side note, I designed the iOS client to drop events if an error occurs on network send, because the network connection automatically retries, so if it fails, it's failed after retrying. I wasn't sure if that's the case here.

Reply:
Gotcha.  Snowplow JS Tracker handles retrying.  I'll change this code to explicitly drop the events for now.